src/Mod/Fem/femmesh/gmshtools.py

```diff
@@ -198,14 +198,15 @@ def get_dimension(self):
             print("Error in dimension")
         print("  ElementDimension: " + self.dimension)
 
-    def get_tmp_file_paths(self):
+    def get_tmp_file_paths(self, tmpdir=None):
         if system() == "Linux":
             path_sep = "/"
         elif system() == "Windows":
             path_sep = "\\"
         else:
-            path_sep = "/"
-        tmpdir = tempfile.gettempdir()
+            path_sep = "/"        
```

Comment:
Thanks for the quick reply and for this new information. Fantastic!
Maybe this optional tmpdir could be an alternate way of doing it without even having to import os? This seems to be the natural way in https://github.com/ElmerCSC/elmerfem/tree/devel/ElmerWorkflows/FreeCADBatchFEMTools
that we are trying to use. 

Many thanks for your comment and support! ;)

Reply:
I don't think there is any benefit in avoiding use of `os`, as it's a standard module.

Reply:
Except one line less to write and a tiny speedup 😉

What if someone wants to write the gmsh file to a dedicated location and at the same time uses the standard tmp for something else? Then he might be forced to switching back and forth.

Reply:
I'm not sure what you mean. I'm not sure what the code does exactly; I don't really care about the implementation. I was just suggesting that instead of manually defining the path separator, and building manually the path, this could be done with `os`, that's it. You don't have to use a temporary directory at all, or you can use as many other locations as you want, but using `os` is better.

Reply:
I misunderstood you. Sorry! You are completely right.

Reply:
there is one big advantage of using os.path.join. One does not need ot take care if the pathes passed already have a separator or not. If join is not used one ends up in having double separator or no separator. 

Thus I started to replace this separator code in all FEM and use os.path.join. But it seams it is still used on some places.

BTW: great work these library. I would recommend to try the library with 0.19 too. There have been lots of fixes and small improvements in 0.19 FEM already.

We should not use a separate parameter. IMHO we should avoid the problem you have at all. Someone could start a second FreeCAD and runs gmsh in each we would run into trouble. 

I will have a look into this. 

cheers bernd

Reply:
@berndhahnebach Yes, I agree with you both on the os thing. I completely misunderstood what @vocx-fc said. I'm sorry!

On the issue itself. While I agree that it might not be solved completely with what is proposed, IMHO the optional parameter for deciding the dumping place for the "temporary" files might be a good idea in any case. For example what we are now doing is that in our modeling system we run a number of cases in parallel each having their own case id and folder (and their own FreeCAD process). So for us it would be convenient that we could tell the create_mesh() function where we want the stuff to be written (for example in case of debugging).

I appreciate your comment about our library. Thank you very much! Yes we should try to update our FreeCAD and test the library on it.

Thank you for reminding about the forum. Sorry!

Reply:
got your point in the regard of the parameter and makes sense to me too.

bernd

Reply:
In the version you provided you need to overwrite the create_mesh(). How about this? https://github.com/berndhahnebach/FreeCAD_bhb/commit/8b0b259b6138993c3944601491d9897a12aaad52

Reply:
Very impressive. Good Job!  The only thing I'm now thinking is that the tmpdir needs to be always injected somehow to this Gmsh object that still requires some changes. For example in our case our interface is ObjectsFem.makeMeshGmsh. Now the tmpdir needs to be passed through that.

Personally, I don't see any problem in changing the create_mesh() code since it is backwards compatible with optional path. 

I would probably merge your suggestion as it is, very nice job with the uuid btw.! And still keep my commit since from there the tmpdir could be easily changed even later on. Imagine you make two meshes with very little changes and you quickly want to change path.

Reply:
May be we could make it even smarter ...
 related https://forum.freecadweb.org/viewtopic.php?f=18&t=23702&p=331752#p331752  

ahh and just realised there is even some better than os.path.join() ...  with the use from of from pathlib import Path the separeators are all managed by the module. Means even no mixed separators in a path like in os.path

Reply:
<img width="967" alt="screen" src="https://user-images.githubusercontent.com/6129869/64255237-3c0f0b00-cf21-11e9-865f-077d8105e767.png">

Reply:
https://github.com/FreeCAD/FreeCAD/commit/8eadf5b78972cf600bac707190a5d023ac51759b

in addition the solver working dir preferences are used. Means you could save the mesh besides the solver input file